README.md
# prettier [![Build Status](https://travis-ci.org/mishadoff/prettier.svg?branch=master)](https://travis-ci.org/mishadoff/prettier) [![Coverage Status](https://coveralls.io/repos/github/mishadoff/prettier/badge.svg?branch=master)](https://coveralls.io/github/mishadoff/prettier?branch=master)

Library designed to make things look _prettier_.

## Usage

`[com.mishadoff/prettier "0.1.1"]`


[intro.md](https://github.com/mishadoff/prettier/blob/master/doc/intro.md)

## License

**EPLv2.0** 

mishadoff © 2019 


doc/release_notes.md
# 0.1.1 Maybe package added (2019-06-09)

- **prettier.maybe** (candidates for invalid values)

# 0.1.0 Initial Release (2019-04-29)

- **prettier.size** (bytes to human-readable conversion)
- **prettier.time** (milliseconds to human-readable conversion)
- **prettier.diff** (diff between maps with support of EntryAdded, EntryDeleted, ValueChanged and KeyRenamed)


test/prettier/diff_test.clj
(ns prettier.diff-test
  (:require [clojure.test :refer :all]
            [prettier.diff :as diff]))

(deftest maps--basic-added-deleted-test
  (is (= [(diff/->MapEntryAdded [:a] 1)] (diff/maps {} {:a 1})))
  (is (= [(diff/->MapEntryDeleted [:a] 1)] (diff/maps {:a 1} {})))
  (is (= [(diff/->MapEntryAdded [:a] 1)
          (diff/->MapEntryAdded [:b] 2)
          (diff/->MapEntryAdded [:c] 3)] (diff/maps {} {:a 1 :b 2 :c 3})))
  (is (= [(diff/->MapEntryDeleted [:a] 1)
          (diff/->MapEntryAdded [:b] 2)] (diff/maps {:a 1} {:b 2}))))

(deftest maps--nested-added-deleted-test
  (is (= [(diff/->MapEntryAdded [:a :b :c] 1)] (diff/maps {} {:a {:b {:c 1}}})))
  (is (= [(diff/->MapEntryAdded [:d] 2)
          (diff/->MapEntryAdded [:a :b :c] 1)] (diff/maps {} {:a {:b {:c 1}} :d 2})))
  (is (= [(diff/->MapEntryDeleted [:a :b :d] 1)
          (diff/->MapEntryAdded [:a :b :c] 2)] (diff/maps {:a {:b {:d 1}}} {:a {:b {:c 2}}}))))

(deftest maps--basic-edited-test
  (is (= [(diff/->MapValueEdited [:a] 1 2)] (diff/maps {:a 1} {:a 2})))
  (is (= [(diff/->MapValueEdited [:a :b] 1 2)] (diff/maps {:a {:b 1}} {:a {:b 2}})))
  (is (= [(diff/->MapValueEdited [:c] 1 2)
          (diff/->MapValueEdited [:a :b] 1 2)] (diff/maps {:a {:b 1} :c 1} {:a {:b 2} :c 2}))))

(deftest maps--basic-renamed-test
  (is (= [(diff/->MapKeyRenamed [:a] [:b] 1)] (diff/maps {:a 1} {:b 1})))
  (is (= [(diff/->MapKeyRenamed [:a :b] [:b] 1)] (diff/maps {:a {:b 1}} {:b 1})))
  (is (= [] (diff/maps {:a {:b 1}} {:a {:b 1}}))))

test/prettier/size_test.clj
(ns prettier.size-test
  (:require [clojure.test :refer :all]
            [prettier.size :as size]))

(deftest bytes->readable--defaults-test
  (is (= "0 B" (size/bytes->readable 0)))
  (is (= "1000 B" (size/bytes->readable 1000)))
  (is (= "1.0 KB" (size/bytes->readable 1024)))
  (is (= "1.2 KB" (size/bytes->readable 1224)))
  (is (= "1.0 MB" (size/bytes->readable (* 1024 1024))))
  (is (= "2.5 GB" (size/bytes->readable 2732746123)))
  (is (= "7.1 TB" (size/bytes->readable 7823645782369)))
  (is (= "1.5 PB" (size/bytes->readable 1726387216472146)))
  (is (= "8.9 PB" (size/bytes->readable 10000000000000000)))
  (is (= "8.7 EB" (size/bytes->readable 10000000000000000000)))
  (is (= "8.5 ZB" (size/bytes->readable 10000000000000000000000)))
  (is (= "8.3 YB" (size/bytes->readable 10000000000000000000000000)))
  (is (= "Infinity" (size/bytes->readable 10000000000000000000000000000))))

(deftest bytes->readable--errors-test
  (is (thrown? AssertionError (size/bytes->readable -1)))
  (is (thrown? AssertionError (size/bytes->readable -1024)))
  (is (thrown? AssertionError (size/bytes->readable "ef")))
  (is (thrown? AssertionError (size/bytes->readable {:a [1 2]}))))

(deftest bytes->readable--customizations-test
  (binding [size/*size-abbreviations* ["bytes" "kilobytes" "megabytes" "gigabytes"]
            size/*size-decimal-format* "%.0f"
            size/*size-over-limit* "...too much..."
            size/*size-gap* ""]
    (is (= "5bytes" (size/bytes->readable 5)))
    (is (= "22kilobytes" (size/bytes->readable 23000)))
    (is (= "3megabytes" (size/bytes->readable 2723446)))
    (is (= "3gigabytes" (size/bytes->readable 3723446000)))
    (is (= "...too much..." (size/bytes->readable 1723446000000)))))

(deftest bytes->readable--classic-units-test
  (binding [size/*power* 1000]
    (is (= "500 B" (size/bytes->readable 500)))
    (is (= "1.0 KB" (size/bytes->readable 1000)))
    (is (= "5.0 MB" (size/bytes->readable (* 5 1000 1000))))))

;;;

(deftest readable->bytes--test
  (is (= 1 (size/readable->bytes "1b")))
  (is (= 0 (size/readable->bytes "0b")))
  (is (= 1024 (size/readable->bytes "1kb")))
  (is (= 1024 (size/readable->bytes "1 KB")))
  (is (= 102400 (size/readable->bytes "100K")))
  (is (= 104857600 (size/readable->bytes "100M")))
  (is (= 524288000 (size/readable->bytes "500mb")))
  (is (= 5368709120 (size/readable->bytes "5G")))
  (is (= 1319413953331 (size/readable->bytes "1.2Tb")))
  (is (= 2417 (size/readable->bytes "0.000000000000000000002YB"))))

(deftest readable->bytes--errors-test
  (is (nil? (size/readable->bytes "")))
  (is (nil? (size/readable->bytes "100")))
  (is (nil? (size/readable->bytes "GB")))
  (is (nil? (size/readable->bytes "1 Megabyte")))
  (is (nil? (size/readable->bytes "-100K")))
  (is (thrown? AssertionError (size/readable->bytes 1)))
  (is (thrown? AssertionError (size/readable->bytes nil)))
  (is (thrown? AssertionError (size/readable->bytes {}))))

(deftest readable->bytes--customizations-test
  (binding [size/*size-parseable-units* [#"(?i)byte" #"(?i)kilobyte" #"(?i)megabyte"]]
    (is (= 1 (size/readable->bytes "1 Byte")))
    (is (= 1024 (size/readable->bytes "1 Kilobyte")))
    (is (= 1048576 (size/readable->bytes "1 Megabyte")))
    (is (nil? (size/readable->bytes "1 Gigabyte")))))

test/prettier/time_test.clj
(ns prettier.time-test
  (:require [clojure.test :refer :all]
            [prettier.time :as time]))

(deftest interval->readable--defaults-test
  (is (= "123 milliseconds" (time/ms->readable 123)))
  (is (= "2 seconds 100 milliseconds" (time/ms->readable 2100)))
  (is (= "5 minutes" (time/ms->readable 300000)))
  (is (= "4 hours 10 minutes 200 milliseconds" (time/ms->readable 15000200)))
  (is (= "11 days 8 hours 13 minutes 20 seconds" (time/ms->readable 980000000)))
  (is (= "3 months 25 days 17 hours 46 minutes 40 seconds" (time/ms->readable 10000000000)))
  (is (= "32 years 1 months 24 days 1 hours 46 minutes 40 seconds" (time/ms->readable 1000000000000))))

(deftest ms->readable--single-unit-test
  (binding [time/*time-units* {:hours "hours"}]
    (is (= "79.5 hours" (time/ms->readable 286217846)))
    (is (= "0.0 hours" (time/ms->readable 100000))))
  (binding [time/*time-units* {:years "ys"}
            time/*time-decimal-format* "%.6f"]
    (is (= "0.000032 ys" (time/ms->readable 1000000)))
    (is (= "321.502058 ys" (time/ms->readable 10000000000000)))))

(deftest ms->readable--errors-test
  (is (thrown? AssertionError (time/ms->readable "time")))
  (is (thrown? AssertionError (time/ms->readable {:map 1})))
  (is (thrown? AssertionError (time/ms->readable 0)))
  (is (thrown? AssertionError (time/ms->readable -10000))))

(deftest ms->readable--custom-labels-test
  (binding [time/*time-units* {:seconds      "s"
                               :minutes      "m"
                               :hours        "h"
                               :days         "d"}
            time/*time-unit-gap* ""
            time/*time-unit-separator* ""]
    (is (= "1m40s" (time/ms->readable 100000)))
    (is (= "2h28m54s" (time/ms->readable 8934751)))
    (is (= "8d11h32m36s" (time/ms->readable 732756812)))))

(deftest ms->readable--custom-time-hierarchy-test
  (binding [time/*time-hierarchy* [[:days (* 1000 60 60 24)]
                                   [:weeks 7]
                                   [:months 4]]
            time/*time-units* {:days "DAYS" :weeks "WEEKS" :months "MONTHS"}]
    (is (= "2 MONTHS 1 WEEKS 5 DAYS" (time/ms->readable 5900000000)))
    (is (= "3 WEEKS 2 DAYS" (time/ms->readable 2000000000)))))

(deftest ms->readable--invalid-time-units
  (binding [time/*time-units* {}]
    (is (thrown? AssertionError (time/ms->readable 1)))))

;;;

(deftest readable->ms--success-test
  (is (= 0 (time/readable->ms "0ms")))
  (is (= 132 (time/readable->ms "132 ms")))
  (is (= 5 (time/readable->ms "5 milliseconds")))
  (is (= 30000 (time/readable->ms "30 seconds")))
  (is (= 180000 (time/readable->ms "3m")))
  (is (= 300000 (time/readable->ms "5 mins")))
  (is (= 300000 (time/readable->ms "5 MINUTES")))
  (is (= 300000 (time/readable->ms "5MINUTES")))
  (is (= 5400000 (time/readable->ms "1.5 hours")))
  (is (= 234316800 (time/readable->ms "2.712 days")))
  (is (= 7776000000 (time/readable->ms "3 months")))
  (is (= 31104000000N (time/readable->ms "1y"))))

(deftest readable->ms--customization-test
  (binding [time/*reverse-time-units* {:seconds #{"секунд" "секунди" "секунда"}}]
    (is (= 30000 (time/readable->ms "30 СЕКУНД")))
    (is (= 28000 (time/readable->ms "28 секунд")))
    (is (= 2000 (time/readable->ms "2 секунди")))
    (is (= 21000 (time/readable->ms "21 секунда")))
    (is (= 2800 (time/readable->ms "2.8 секунд")))))

(deftest readable->ms-errors-test
  (is (nil? (time/readable->ms "1")))
  (is (nil? (time/readable->ms "1DAYZ")))
  (is (nil? (time/readable->ms "1.2.3"))))

(deftest readable->ms-ambiguity-test
  (binding [time/*reverse-time-units* {:minutes #{"m"}
                                       :months  #{"m" "months"}}]
    (is (thrown? AssertionError (time/readable->ms "3m")))))

src/prettier/diff.clj
(ns prettier.diff)

;; Registered changes
(defrecord MapEntryAdded [key value])
(defrecord MapEntryDeleted [key value])
(defrecord MapValueEdited [key value-from value-to])
(defrecord MapKeyRenamed [key-from key-to value])

(defn- map-into-kv-pairs
  "Transforms nested map into key value pairs. Return vec of pairs
   Each pair represented as vec [key-path value]"
  [m]
  (loop [[k & ks] (->> m keys (map vector)) pairs []]
    (let [v (get-in m k)]
      (cond (nil? k) pairs
            (map? v) (recur (->> (keys v)
                                 (map (partial conj k))
                                 (concat ks))
                            pairs)
            :else (recur ks (conj pairs [k v]))))))

;;;

(defn- maps-basic-changes
  "Detect basic changes between unnested maps
   Handles MapEntryAdded and MapEntryDelete types."
  [before after change-fn]
  (->> before
       (map (fn [[k v]]
              (when (not= v (get after k))
                (change-fn k v))))
       (remove nil?)))

;;;

(defn maps-edited-changes
  "Detects 'edited' changes using basic 'added' and 'deleted' changes
   Uses squash rule to merge two changes into one
    MapEntryDeleted{:key [:a], :value 1}
    MapEntryAdded{:key [:a], :value 2}
                vvv
    MapValueChanged{:key [:a] :value-from 1 :value-to 2}

   Returns vector of 2 values
    - edited changes to be added to result
    - added and deleted changes used for merge to be deleted from result"
  [basic-changes]
  (->> basic-changes
       (group-by :key)
       (filter (fn [[_ changes]]
                 (and (= (count changes) 2)
                      (= #{MapEntryAdded MapEntryDeleted}
                         (->> changes (map type) (into #{}))))))
       (map (fn [[k changes]]
              (let [deleted (->> changes (filter #(= MapEntryDeleted (type %))) first)
                    added (->> changes (filter #(= MapEntryAdded (type %))) first)]
                [(->MapValueEdited k (:value deleted) (:value added)) changes])))
       (reduce (fn [[acc-e acc-changes] [e changes]]
                 [(conj acc-e e) (concat changes acc-changes)])
               [[] []])))

;;;

(defn maps-renamed-changes
  "Detects 'renamed' changes using basic :added and :deleted changes
   Uses squash rule to merge two changes into one
    MapEntryDeleted{:key [:a], :value 1}
      MapEntryAdded{:key [:b], :value 1}
                vvv
    MapKeyRenamed{:key-from [:a], :key-to [:b], :value 1}

   Returns vector of 2 values
    - renamed changes to be added to result
    - added and deleted changes used for merge to be deleted from result"
  [basic-changes]
  (->> basic-changes
       (group-by :value)
       (filter (fn [[_ changes]]
                 (and (= (count changes) 2)
                      (= #{MapEntryAdded MapEntryDeleted}
                         (->> changes (map type) (into #{}))))))
       (map (fn [[v changes]]
              (let [deleted (->> changes (filter #(= MapEntryDeleted (type %))) first)
                    added (->> changes (filter #(= MapEntryAdded (type %))) first)]
                [(->MapKeyRenamed (:key deleted) (:key added) v) changes])))
       (reduce (fn [[acc-e acc-changes] [e changes]]
                 [(conj acc-e e) (concat changes acc-changes)])
               [[] []])))

;;;

(defn maps
  "Find a difference between two maps.
   Returns a sequence of changes, which if applied in order
   transform map 'before' into map 'after'
   Maps could be nested.

   Examples of basic changes:
    MapEntryDeleted{:key [:a], :value 1}
    MapEntryAdded{:key [:a :b :c], :value 10}

   Can squash basic changes to detect more advanced changes:
    MapValueEdited{:key [:a], :value-from 1, :value-to 2}
    MapValueRenamed{:key-from [:a], :key-to [:b], :value 1}
   "
  [before after]
  (let [b-pairs (->> before map-into-kv-pairs (into {}))
        a-pairs (->> after map-into-kv-pairs (into {}))
        deleted (maps-basic-changes b-pairs a-pairs ->MapEntryDeleted)
        added (maps-basic-changes a-pairs b-pairs ->MapEntryAdded)
        added-and-deleted (concat deleted added)
        enrich (fn [changes changes-fn]
                 (let [[to-add to-remove] (changes-fn changes)
                       to-remove-set (into #{} to-remove)]
                   (concat (remove to-remove-set changes) to-add)))]
    (-> added-and-deleted
        (enrich maps-edited-changes)
        (enrich maps-renamed-changes))))

src/prettier/size.clj
(ns prettier.size
  (:require [prettier.util :as u]))

(def ^:dynamic *size-abbreviations*
  ["B" "KB" "MB" "GB" "TB" "PB" "EB" "ZB" "YB"])

(def ^:dynamic *power* 1024)

(def ^:dynamic *size-decimal-format* "%.1f")
(def ^:dynamic *size-gap* " ")
(def ^:dynamic *size-over-limit* "Infinity")

;;;

(defn bytes->readable
  "Transform integer number of bytes into human readable format"
  [b]
  {:pre [(and (integer? b) (>= b 0))]}
  (loop [order 0 bnum (* 1.0 b)]
    (cond
      (>= order (count *size-abbreviations*)) *size-over-limit*
      (<= 0 bnum (dec *power*))
      (let [effective-decimal-fmt (if (zero? order) "%.0f" *size-decimal-format*)]
            (format (str effective-decimal-fmt *size-gap* "%s")
                    bnum
                    (nth *size-abbreviations* order)))
      :else (recur (inc order) (/ bnum *power*)))))

;;;

(def ^:dynamic *size-parseable-units*
  [#"(?i)b"
   #"(?i)k|kb"
   #"(?i)m|mb"
   #"(?i)g|gb"
   #"(?i)t|tb"
   #"(?i)p|pb"
   #"(?i)e|eb"
   #"(?i)z|zb"
   #"(?i)y|yb"])

(defn readable->bytes
  "Transform readable size string into bytes integer"
  [s]
  {:pre [(string? s)]}
  (let [[num unit] (u/number-and-unit s)
        order (->> *size-parseable-units*
                   (map-indexed (fn [idx re]
                                  [(some->> unit
                                            (re-matches re))
                                   idx]))
                   (filter first)
                   first
                   second)]
    (when (and num (>= num 0) order)
      (-> num
          (* (->> (repeat order *power*)
                  (reduce *')))
          (bigint)))))

src/prettier/util.clj
(ns prettier.util
  (:require [clojure.string :as str]))

(defn number-and-unit
  "Parse a string and returns number and unit.
   Both number and unit is required, nil otherwise"
  [s]
  (let [[_ p1 p2 unit] (re-matches #"(\-?\d+)(\.\d+)?\s*(.*)" s)]
    (if-not (or (str/blank? p1) (str/blank? unit))
      [(-> (str p1 p2) (bigdec))
       unit])))

(defn distance
  "Calculates levenshtein distance between two sequences
  Levenshtein distance roughly defines number of changes
  which required to apply to transform 'seq1' into 'seq2'
  Example: consider strings as sequence of letters
  (distance \"bear\" \"beer\") => 1 (1st change from 'a' to 'e')
  (distance \"dog\" \"doll\") => 2 (1st change from 'g' to 'l', 2nd change insertion of 'l')"
  [seq1 seq2]
  (let [seq1v (into [] seq1)
        get (fn [v i] (get v i ::not-found))
        init-row (into [] (range (inc (count seq1))))
        advance-row (fn [row e]
                      (reduce (fn [acc i]
                                (cond (= 0 i) (conj acc (inc (get row i)))
                                      (= e (get seq1v (dec i))) (conj acc (get row (dec i)))
                                      :else (conj acc
                                                  (inc (min (get row (dec i))
                                                            (get row i)
                                                            (peek acc))))))
                              []
                              init-row))]
    (peek (reduce advance-row init-row seq2))))

(defn duplicates
  "Return list of duplicates in collection.
   Duplicate is an item which appears more than once"
  [coll]
  (->> coll
       (frequencies)
       (filter (comp (partial < 1) second))
       (map first)
       (into #{})))

src/prettier/time.clj
(ns prettier.time
  (:require [clojure.string :as str]
            [prettier.util :as u]))

(def ^:dynamic *time-hierarchy*
  [[:milliseconds 1]
   [:seconds 1000]
   [:minutes 60]
   [:hours 60]
   [:days 24]
   [:months 30]
   [:years 12]])

;; If provided only one time unit use decimal
(def ^:dynamic *time-units*
  {:milliseconds "milliseconds"
   :seconds      "seconds"
   :minutes      "minutes"
   :hours        "hours"
   :days         "days"
   :months       "months"
   :years        "years"})

(def ^:dynamic *reverse-time-units*
  {:milliseconds #{"ms" "msec" "msecs" "millisecond" "milliseconds"}
   :seconds #{"s" "sec" "secs" "second" "seconds"}
   :minutes #{"m" "min" "mins" "minute" "minutes"}
   :hours #{"h" "hs" "hour" "hours"}
   :days #{"d" "ds" "day" "days"}
   :months #{"mon" "month" "months"}
   :years #{"y" "ys" "year" "years"}})

(def ^:dynamic *time-unit-gap* " ")
(def ^:dynamic *time-unit-separator* " ")
(def ^:dynamic *time-decimal-format* "%.1f")

(defn- time-hierarchy-reductions []
  (reductions
    (fn [[_ v1] [u2 v2]]
      [u2 (* v1 v2)]) *time-hierarchy*))

;;;

(defn ms->readable [ms]
  {:pre [(and (not (empty? *time-units*))
              (integer? ms)
              (> ms 0))]}
  (let [effective-time-units
        (->> (time-hierarchy-reductions)
             (filter (fn [[t _]] (*time-units* t)))
             (map (fn [[t _ :as v]] (conj v (*time-units* t))))
             (reverse))]
    ;; If only one time unit requested, show decimals
    (if (= 1 (count effective-time-units))
      (let [[t amount] (first effective-time-units)]
        (-> (format *time-decimal-format* (/ (* 1.0 ms) amount))
            (str *time-unit-gap* (*time-units* t))))
      ;; Calculate [unit amount] chunks
      (loop [[unit & units] effective-time-units num ms chunks []]
        (cond
          unit
          (let [[t amount] unit
                [p1 p2] ((juxt quot rem) num amount)]
            (recur units p2 (conj chunks [t p1])))
          :else (->> chunks
                     (filter (fn [[_ v]] (> v 0)))
                     (map (fn [[t v]] (str v *time-unit-gap* (*time-units* t))))
                     (str/join *time-unit-separator*)))))))

(defn readable->ms [s]
  (let [[num unit] (u/number-and-unit s)
        unit (some-> unit str/lower-case)
        known-time-units (->> (time-hierarchy-reductions)
                              (into {}))]
    (some->> *reverse-time-units*
             (map (fn [[unit-key candidates]]
                    (when (candidates unit) unit-key)))
             (remove nil?)
             ;; ambiguity check
             ((fn [v]
                (if (> (count v) 1)
                  (throw (AssertionError. (str "Ambiguity for " '*reverse-time-units* ":" v)))
                  v)))
             (first)
             (known-time-units)
             (* num)
             (bigint))))